hamlpy/attribute_dict_parser.py

```diff
@@ -0,0 +1,213 @@
+from __future__ import print_function, unicode_literals
+
+import re
+
+from collections import OrderedDict
+
+# Valid characters for dictionary key
+re_key = re.compile(r'[a-zA-Z0-9-_]+')
+re_nums = re.compile(r'[0-9\.]+')
+re_whitespace = re.compile(r'([ \t]+)')
+re_leading_spaces = re.compile(r'^\s+', re.MULTILINE)
+re_line = re.compile(r'.*')
+re_sq = re.compile(r'\'([^\'\\]|\\.)*\'')
+re_dq = re.compile(r'\"([^\"\\]|\\.)*\"')
+
+class AttributeParser:
+    """Parses comma-separated HamlPy attribute values"""
+
+    def __init__(self, data, terminator):
+        self.terminator=terminator
+        self.s = data.lstrip()
+        self.length=len(self.s)
+        # Index of current character being read
+        self.ptr=1
+
+    def consume_whitespace(self, include_newlines=False):
+        """Moves the pointer to the next non-whitespace character"""
+        whitespace = (' ', '\t', '\r', '\n') if include_newlines else (' ', '\t')
+
+        while self.ptr<self.length and self.s[self.ptr] in whitespace:
```

Comment:
I think on line 111 the reason for `self.length-1` is that it's parsing up to but not including the final `}`